### PR TITLE
SILCombine: peephole to propagate resilient enum cases

### PR DESCRIPTION
**Pull request by eeckstein**
Basically the pattern to optimize is:
    inject_enum_addr %stackloc, #SomeCase
    switch_enum_addr %stackloc ...
    
This works even if the enum is resilient or not a loadable type (e.g. an enum with a not loadable payload) and the case does not have a payload. As long as we don't have opaque values in SIL we need this peephole to optimize the pattern.
This change fixes the code generation for Float.rounded().
    
rdar://problem/46353885

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by slavapestov** [comment]
This is not specific to resilient enums is it? We use the same SIL instructions for any enum with address only representation, for example if you have an existential payload.

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp [hunk 1]
```diff
@@ -204,6 +205,23 @@ static bool canZapInstruction(SILInstruction *Inst, bool acceptRefCountInsts) {
       isa<DeallocPartialRefInst>(Inst))
     return acceptRefCountInsts;
 
+  if (isa<InjectEnumAddrInst>(Inst))
+    return true;
+
+  // We know that the destructor has no side effects so we can remove the
+  // deallocation instruction too.
+  if (isa<DeallocationInst>(Inst) || isa<AllocationInst>(Inst))
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: I noticed that you are checking AllocationInst here as well. That is new. Why are you changing that?

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp [hunk 2]
```diff
@@ -204,6 +205,23 @@ static bool canZapInstruction(SILInstruction *Inst, bool acceptRefCountInsts) {
...
+    return true;
+
+  // Much like deallocation, destroy addr is safe.
+  if (isa<DestroyAddrInst>(Inst))
+    return true;
+
+  // The only store instructions which is guaranteed to store a trivial value
+  // is an inject_enum_addr witout a payload (i.e. without init_enum_data_addr).
+  if (onlyAcceptTrivialStores)
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: I find onlyAcceptTrivialStores to be confusing slightly. I think what you are really trying to say that the value is non-trivial, so only inject_enum_addr without a payload is the only way to have a trivial store. Can you make that clearer? I had to think about this to make sure it made sense to me. That says to me that it needs a comment or to be written clearer.

**Review by gottesmm — CHANGES_REQUESTED** [review]
Overall the functionality looks ok to me. Please see my comments.

**Comment by eeckstein** [comment]
> This is not specific to resilient enums is it?

You are right, my title/description is a bit too narrow.

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: It's needed to remove dead alloc_stack/dealloc_stack of non-trivial types.

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp (hunk 2 above)
Comment (eeckstein) [reply lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: That's what I already wrote in that comment. If you have a specific suggestion for a better comment, please let me know

File: lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp [hunk 3]
```diff
@@ -106,23 +106,91 @@ SILCombiner::visitAllocExistentialBoxInst(AllocExistentialBoxInst *AEBI) {
...
+          return nullptr;
+      }
+    }
+    if (NumWrites != 1)
+      return nullptr;
+    if (auto *IEA = dyn_cast<InjectEnumAddrInst>(WritingInst))
+      return IEA;
+
+    // In case of a copy_addr continue with the source of the copy.
+    Addr = dyn_cast<CopyAddrInst>(WritingInst)->getSrc();
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp]: ok, I changed it

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: How did we use to handle that before? Did we just leave them in?

File: lib/SILOptimizer/Transforms/DeadObjectElimination.cpp (hunk 2 above)
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/DeadObjectElimination.cpp]: I would add into the comment that this applies only to non-trivial types. The reason I was confused is b/c you can store [trivial]. That is why it was confusing to me.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Comment by gottesmm** [comment]
LGTM

**Comment by shahmishal** [comment]
@swift-ci test macOS

**Review by gottesmm — APPROVED** [review]

**Pull request merged**